File: 1236-n-th-tribonacci-number/n-th-tribonacci-number.cpp


class Solution 
{
public:
    int tribonacci( int n ) 
    {
        if( n == 0 ) return 0;
        if( n == 1 ) return 1;
        if( n == 2 ) return 1;

        vector<int> table( n + 1 );
        table[ 0 ] = 0;
        table[ 1 ] = 1;
        table[ 2 ] = 1;
        
        for( int i = 3; i <= n; i++ )
        {
            table[ i ] = table[ i - 1 ] + table[ i - 2 ] + table[ i - 3 ];
        }
        
        return table[ n ] ;
    }
};

File: 1236-n-th-tribonacci-number/n-th-tribonacci-number.py
class Solution:
    
    d = {}

    def tribonacci( self, n: int ) -> int:
        if n == 0: return 0
        if n == 1: return 1
        if n == 2: return 1
        if n in self.d.keys(): return self.d[ n ]

        self.d[ n ] = sum([ self.tribonacci( n - i ) for i in range( 1, 4 )])

        return self.d[ n ]
        